### DISCUSSION TITLE
How to just validate or assert, not parse, with zod?

### DISCUSSION THREAD

**Discussion by lancejpollard** [colinhacks/zod#2450] [Q&A]
Is there a way to just do an assertion or validation with Zod, and not have it construct a new object? I would like to do this:

```ts
function assertSomething(input: object) {
  userSchema.assert(input)
  input.email // ...
}

// not asserts, but just x is type:
function checkSomething(input: object) {
  if (userSchema.is(input)) {
    input.email // ...
  }
}
```

Is that possible, or must I always construct a new object?

Perhaps all it takes is something like this?

```ts
export function assertType<T, S extends z.ZodType<T>>(
  x: unknown,
  schema: S,
): asserts x is T {
  schema.parse(x)
}

export function isType<T, S extends z.ZodType<T>>(
  x: unknown,
  schema: S,
): x is T {
  return schema.safeParse(x).success
}
```

How would I properly write those 2 functions if so?

**Comment by scotttrinh** [comment]
Be careful here! Zod allows transformations, which is why this is not just a simple built-in concept in Zod's core.

There are many drawbacks to making this an assertion function, and very few advantages, so before you travel down this path, just make sure you are extremely careful. Let me make a quick example:

```ts
const Data = z.object({
  foo: z.coerce.number(),
  bar: z.boolean.transform((v) => v ? "true" : "false"),
});

const data = {
  foo: "123",
  bar: false,
};

Data.assert(data);
data.foo * 10// oh no, this is still the string "123", but the type is `number`!
data.bar ? launchMissiles() : rest() // oh no, this is the string `"false"` so it will launch missiles since a non-empty string is truthy
```

What you end up having to do is have separate assertions for input types and output types and they're _still_ wrong if you use `preprocess` or `coerce`.
